### PR TITLE
drivers: ieee802154: support Key Identifier Mode > 1

### PR DESCRIPTION
**Pull request by jciupis**
IEEE 802.15.4-2020 defines four possible values for Key Identifier Mode field of the Auxiliary Security Header. The current ieee802154 driver API only supports two of them: b00 and b01. This commit adds support for the two remaining Key Identifier Mode values. It's done by replacing a field that can only hold Key Index into a field that can holds a pointer to the entire Key Identifier field.

See IEEE 802.15.4-2020, sections 9.4.2.3 and 9.4.4 for further reference.

### CODE REVIEW

**Review by edmont — DISMISSED** [review]

**Comment by jciupis** [comment]
FYI: @ankuns

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 1]
```diff
@@ -781,7 +781,7 @@ static void nrf5_config_mac_keys(struct ieee802154_key *mac_keys)
 		nrf_802154_key_t key = {
 			.value.p_cleartext_key = keys->key_value,
 			.id.mode = keys->key_id_mode,
-			.id.p_key_id = &(keys->key_index),
+			.id.p_key_id = keys->key_id,
```
Comment (ankuns) [inline drivers/ieee802154/ieee802154_nrf5.c:784]: Is it enough? The meaning of what `key.id.p_key_id` points to is changed. I can point to multi-byte array.
What about line 796? `stored_ids` was also prepared for storing old `key_index` of `uint8_t` type.
Also line 798 seems now incorrect. stored_key_ids[i].p_key_id will point now to one byte in stored_ids[i] no matter what's the size of real key_id.

**Review by canisLupus1313 — DISMISSED** [review]

File: drivers/ieee802154/ieee802154_nrf5.c (hunk 1 above)
Comment (jciupis) [reply drivers/ieee802154/ieee802154_nrf5.c:784]: That's a valid concern, thanks for catching it. I put `DNM` label on the PR and will come back with a solution next week.

Comment (ankuns) [reply drivers/ieee802154/ieee802154_nrf5.c:784]: All this issues come from the fact that there is no function `nrf_802154_security_key_remove_all(void)`.
To be able to remove all keys we stored we need to track the key ids we have written.

**Review by canisLupus1313 — DISMISSED** [review]

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 2]
```diff
@@ -781,36 +781,24 @@ static void nrf5_iface_init(struct net_if *iface)
 #if defined(CONFIG_NRF_802154_ENCRYPTION)
 static void nrf5_config_mac_keys(struct ieee802154_key *mac_keys)
 {
-	static nrf_802154_key_id_t stored_key_ids[NRF_802154_SECURITY_KEY_STORAGE_SIZE];
-	static uint8_t stored_ids[NRF_802154_SECURITY_KEY_STORAGE_SIZE];
-	uint8_t i;
+	nrf_802154_security_key_remove_all();
```
Comment (edmont) [inline drivers/ieee802154/ieee802154_nrf5.c:784]: Much better approach, thanks.

**Review by edmont — DISMISSED** [review]

**Review by ankuns — APPROVED** [review]

**Review by rlubos — DISMISSED** [review]

File: west.yml [hunk 3]
```diff
@@ -183,7 +183,7 @@ manifest:
       groups:
         - hal
     - name: hal_nordic
-      revision: 56e0b052dff311c2f8eb08c6804e60fc79feb56f
+      revision: pull/153/head
```
Comment (rlubos) [inline west.yml:186]: `6efb0fdef01737fac03ea500b28b6c4d12248af4`

**Review by rlubos — DISMISSED** [review]

**Review by ankuns — APPROVED** [review]

**Comment by jciupis** [comment]
The regression mentioned in the previous comment was caused by the assumption I made that the OpenThread stack always configures ieee802154 mac keys with Key Id Mode == 1. That was wrong because RCP uses Key Id Mode == 0 when resetting. I corrected the implementation and resolved conflicts that came up since last week.

**Comment by jciupis** [comment]
FYI: @canisLupus1313, @edmont

**Review by canisLupus1313 — APPROVED** [review]

**Review by edmont — APPROVED** [review]

**Pull request merged**